locallib.php

```diff
@@ -903,46 +903,39 @@ public function take_from_student($mformdata) {
         return true;
     }
 
-    public function take_from_form_data($formdata) {
+    public function take_attendance($formdata) {
         global $DB, $USER;
-        // TODO: WARNING - $formdata is unclean - comes from direct $_POST - ideally needs a rewrite but we do some cleaning below.
+
+        if (isset($formdata->takenby) && !empty($formdata->takenby)) {
+            $takenby = $formdata->takenby;
+        } else {
+            $takenby = $USER->id;
+        }
         $statuses = implode(',', array_keys( (array)$this->get_statuses() ));
         $now = time();
-        $sesslog = array();
-        $formdata = (array)$formdata;
-        foreach ($formdata as $key => $value) {
-            if (substr($key, 0, 4) == 'user') {
-                $sid = substr($key, 4);
-                if (!(is_numeric($sid) && is_numeric($value))) { // Sanity check on $sid and $value.
-                     print_error('nonnumericid', 'attendance');
-                }
-                $sesslog[$sid] = new stdClass();
-                $sesslog[$sid]->studentid = $sid; // We check is_numeric on this above.
-                $sesslog[$sid]->statusid = $value; // We check is_numeric on this above.
-                $sesslog[$sid]->statusset = $statuses;
-                $sesslog[$sid]->remarks = array_key_exists('remarks'.$sid, $formdata) ?
-                                                      clean_param($formdata['remarks'.$sid], PARAM_TEXT) : '';
-                $sesslog[$sid]->sessionid = $this->pageparams->sessionid;
-                $sesslog[$sid]->timetaken = $now;
-                $sesslog[$sid]->takenby = $USER->id;
-            }
-        }
-
         $dbsesslog = $this->get_session_log($this->pageparams->sessionid);
-        foreach ($sesslog as $log) {
-            if ($log->statusid) {
-                if (array_key_exists($log->studentid, $dbsesslog)) {
-                    $log->id = $dbsesslog[$log->studentid]->id;
-                    $DB->update_record('attendance_log', $log);
+        foreach ($formdata->users as $log) {
+            $record = new stdClass();
+            $record->studentid = $log['userid']; // We check is_numeric on this above.
+            $record->statusid = $log['statusid']; // We check is_numeric on this above.
+            $record->statusset = $statuses;
+            $record->remarks = clean_param($log['remarks'], PARAM_TEXT);
+            $record->sessionid = $formdata->sessionid;
```

Comment:
is this right? - was previously $this->pageparams->sessionid; If we obtain from the formdata should it be cleaned? (I haven't traced the code)